contracts/compounder/GenericVault.sol

```diff
@@ -0,0 +1,335 @@
+// SPDX-License-Identifier: MIT
+pragma solidity 0.8.11;
+
+import { Ownable } from "@openzeppelin/contracts-0.8/access/Ownable.sol";
+import { SafeERC20 } from "@openzeppelin/contracts-0.8/token/ERC20/utils/SafeERC20.sol";
+import { ERC20 } from "@openzeppelin/contracts-0.8/token/ERC20/ERC20.sol";
+import { IERC20 } from "@openzeppelin/contracts-0.8/token/ERC20/IERC20.sol";
+import { ReentrancyGuard } from "@openzeppelin/contracts-0.8/security/ReentrancyGuard.sol";
+import { IERC4626 } from "../interfaces/IERC4626.sol";
+import { IStrategy } from "../interfaces/IStrategy.sol";
+import { IBasicRewards } from "../interfaces/IBasicRewards.sol";
+
+/**
+ * @title   GenericUnionVault
+ * @author  llama.airforce -> AuraFinance
+ * @notice  Changes:
+ *          - remove withdraw penalty
+ *          - remove platform fee
+ *          - add extra rewards logic
+ */
+contract GenericUnionVault is ERC20, IERC4626, Ownable, ReentrancyGuard {
+    using SafeERC20 for IERC20;
+
+    uint256 public withdrawalPenalty = 100;
+    uint256 public constant MAX_WITHDRAWAL_PENALTY = 150;
+    uint256 public constant FEE_DENOMINATOR = 10000;
+
+    address public immutable underlying;
+    address public strategy;
+
+    address[] public extraRewards;
+
+    event WithdrawalPenaltyUpdated(uint256 _penalty);
+    event Harvest(address indexed _caller, uint256 _value);
+    event CallerIncentiveUpdated(uint256 _incentive);
+    event StrategySet(address indexed _strategy);
+    event ExtraRewardAdded(address indexed _reward);
+    event ExtraRewardCleared(address indexed _reward);
+
+    constructor(address _token)
+        ERC20(
+            string(abi.encodePacked("Staked ", ERC20(_token).name())),
+            string(abi.encodePacked("stk", ERC20(_token).symbol()))
+        )
+    {
+        underlying = _token;
+    }
+
+    /// @notice Updates the withdrawal penalty
+    /// @param _penalty - the amount of the new penalty (in BIPS)
+    function setWithdrawalPenalty(uint256 _penalty) external onlyOwner {
+        require(_penalty <= MAX_WITHDRAWAL_PENALTY);
+        withdrawalPenalty = _penalty;
+        emit WithdrawalPenaltyUpdated(_penalty);
+    }
+
+    /// @notice Set the address of the strategy contract
+    /// @dev Can only be set once
+    /// @param _strategy - address of the strategy contract
+    function setStrategy(address _strategy) external onlyOwner notToZeroAddress(_strategy) {
+        require(strategy == address(0), "Strategy already set");
+        strategy = _strategy;
+        emit StrategySet(_strategy);
+    }
+
+    /// @notice Count of extra rewards
+    function extraRewardsLength() external view returns (uint256) {
+        return extraRewards.length;
+    }
+
+    /// @notice Add extra reward contract
+    /// @param _reward VirtualBalanceRewardPool address
+    /// @return bool success
+    function addExtraReward(address _reward) external onlyOwner notToZeroAddress(_reward) returns (bool) {
```

Comment:
I think this should have a check for duplicates. 

Without using a mapping then a function like this;

`
function isExtraReward(address _reward) public returns(bool found) {
    for(uint i =0; i<extraRewards.length; i++) {
        if(extraRewards[i] == _reward) {
            found = true;
            continue;
        }
    }
}
`

and then update the addReward fn;
`
if (extraRewards.length >= 12 && !isExtraReward(_reward) {
`

or just a simple public mapping that we add/remove from.
`
mapping (address => bool) public isExtraReward);
`

Reply:
#177